### ISSUE TITLE
Idiomatic directory structure

### ISSUE THREAD

**Issue by awdeorio** [eecs280staff/tutorials#16]
Modify the directory structure in the setup tutorials to use a subdirectory for source code.

## Example
The current directory structure you get after completing the setup tutorial for project 1 is flat.  The source code is in the top level directory.
```console
$ tree -a p1-stats
p1-stats
├── .gitignore
├── Makefile
├── README.md
├── main.cpp
├── main_test.in
├── main_test.out.correct
├── main_test_data.tsv
├── p1_library.cpp
├── p1_library.h
├── stats.cpp
├── stats.h
├── stats_public_test.cpp
├── stats_tests.cpp
└── stats_tests.cpp.starter
```

The proposed directory structure looks like this.  Notice that the source code is in a subdirectory and the `README` is in the top level directory.
```console
$ pwd
/Users/awdeorio/src/eecs280
$ tree -a p1-stats
p1-stats
├── .gitignore
├── README.md
└── p1-stats
    ├── Makefile
    ├── main.cpp
    ├── main_test.in
    ├── main_test.out.correct
    ├── main_test_data.tsv
    ├── p1_library.cpp
    ├── p1_library.h
    ├── stats.cpp
    ├── stats.h
    ├── stats_public_test.cpp
    ├── stats_tests.cpp
    └── stats_tests.cpp.starter
```

## Motivation
There are several reasons I think this change is a good idea.

1. The idiomatic location for source code is inside a subdirectory.  The top level project directory should contain things like a README or a license.  This is true in C/C++ projects, Python projects, web apps, etc.

2. XCode and Visual Studio create a subdirectory for source code.  We currently use a nasty trick to relocate the files, flattening the directory structure.  It's complicated and weird.

3. EECS 281 uses subdirectories for source code.  The proposed change would bring the 280 and 281 tool chains one step closer together.

4. The EECS 280 tutorials would become a bit more useful outside the course.

## Todo
- [ ] Update XCode tutorial to use default directory structure
- [ ] Update Visual Studio tutorial to use default directory structure
- [ ] Update VS Code tutorial to use a subdirectory
- [ ] Update starter files to use a containing directory
- [ ] Update starter files unpack instruction
- [ ] Search for mentions of `pwd` and check for lingering inconsistencies
- [ ] Create an issue in P2, P3, P4 and P5 repos to update and setup instructions and starter files tarball

**Comment by jamesjuett** [comment]
As I expressed a bit offline and have thought about more, I think the right approach is:

- Make changes to simplify the Visual Studio and Xcode tutorials and play nicely with their project structure.
- Retain the flat directory structure for VS Code.

The main reasoning is:
- The proposed directory structure is not idiomatic. It's simply a flat structure (as we have already) in an unnecessary subdirectory. It's fine to have that subdirectory where IDEs want it and where it will make students lives easier (in VS and XCode), but we should not have it where the IDE doesn't want it and where it will make students lives harder (VS Code).

For the motivations given above in the issue description, I want to provide a few counterpoints:

> 1. The idiomatic location for source code is inside a subdirectory. The top level project directory should contain things like a README or a license. This is true in C/C++ projects, Python projects, web apps, etc.

This may be true if your intent is to put only source code in a subdirectory. But any of the proposals on the table that I'm aware of for our tutorials would put *almost everything* in that subdirectory, including the `Makefile`, compiled binaries, input/data files, etc.. That's not idiomatic in any language, unless you've decided to go with a flat directory structure (in which case it wouldn't be a subdirectory, it would just be the directory). A flat directory structure also seems pretty reasonable for: 1) an educational context 2) a more agnostic public-facing tutorial that doesn't commit to an "opinionated" project config/structure

2. 

> XCode and Visual Studio create a subdirectory for source code. We currently use a nasty trick to relocate the files, flattening the directory structure. It's complicated and weird.

YES! Fair point, and it is worth doing to simplify those tutorials for students. But I think this is the only motivation and it is best done as part of IDE setup for those IDEs. It's also OK if the IDE tutorials have different setups.

> 3. EECS 281 uses subdirectories for source code. The proposed change would bring the 280 and 281 tool chains one step closer together.

Consistency makes sense for Visual Studio and XCode and I think those changes would be helpful for students. Especially given point 2) above. However, I'm fairly certain EECS 281 doesn't provide official support for VS code and encourages students not to use it. It isn't surprising that they use an approach that is preferred by XCode and Visual Studio and that doesn't work well for VS Code. We should use the best approach for each IDE.

> 4. The EECS 280 tutorials would become a bit more useful outside the course.

This will be true for Visual Studio and XCode by removing the weird hack. But the VS Code tutorial will now have a weird hack to put everything in an unnecessary subdirectory.

**Comment by awdeorio** [comment]
I think we agree on these points:
- Students in EECS 280 should put code, tests, Makefile in the same directory
- Students in EECS 281 put code, tests, and Makefile in the same directory
- EECS 281 uses a two-level directory structure (I'll add that EECS 183 does too)
- A two-level directory structure is simpler in Xcode and Visual Studio
- A one-level directory structure is simpler in VS Code

We don't totally agree on these points:
- Magnitude of anticipated confusion when students use a two-level directory structure in VS Code
- Whether the proposed two-level structure is idiomatic enough to call it idiomatic.  Yes, the Makefile and tests would be in the second-level directory.  Other files would not: the README, IDE configuration, gitignore, git metadata, etc.  The location of binaries depends on the IDE, e.g., Xcode and Visual Studio put binaries in a dedicated subdirectory.

A new point you raised: Use a one-level structure in the VS Code tutorial and a two-level structure in the Xcode and Visual Studio tutorials.  I'm open to this idea, but I'm worried about the impact on other tutorials.  In most cases, when we use `pwd`, we mean "change to the directory containing your source code, Makefile, and tests".  I'm worried that we'll end up creating a new class of confusion if some students have a one-level directory structure and other students have a two-level directory structure.

**Comment by jamesjuett** [comment]
(Make sure to read the bottom part where I backtrack!)

**Other Courses**  
I'll add some points about setup in other courses, which are true to the best of my knowledge and add some important context.

ENGR 101  
Supports VS Code and uses a flat directory structure.

EECS 183  
Supports Visual Studio or XCode, but not VS Code. They use a two-level directory structure, but they don't put anything at the top level. So it's _essentially a flat structure, nested within a subdirectory because the IDEs create it that way._

EECS 281  
Supports Visual Studio or XCode using a two-level structure. However, I could not find anything on eecs281.org or the EECS 281 YouTube (where IDE setup videos are hosted) that suggests they put anything at the top level. I'm sure some students in 281 create a README or use git, but I can't find any tutorial material. For all intents and purposes, it appears that they _essentially use a flat structure, nested within a subdirectory because the IDEs create it that way._

It's not clear to me whether 281 "supports" VS Code, but their YouTube channel has a [VS Code setup video](https://www.youtube.com/watch?v=1HGBk78BqR4) that uses a **one-level structure**.

Our other courses generally use a setup that follows naturally from the particular IDE.

**The `pwd` Cue and Consistency**
I don't think most students use this cue anyway. Those that understand it don't really need it, and those that don't understand it just execute `pwd` by rote in their own terminal. This at least is my experience interacting with students in office hours.

I actually think it would be better to have a text cue, perhaps something like you mentioned earlier with "change to the directory containing your source code, Makefile, and tests".

Or, instead of `pwd` they could run `ls` or `ls Makefile` and verify the `Makefile` is present. Or `ls *.cpp` and verify their source is present. For `pwd`, we have to communicate that their path may be different and (we don't really communicate) that you've got to understand that the path is the path to your files (as in your text description).

Either of those approaches (or the current `pwd` cue) might be better supported if we had a quick tutorial for terminal basics that explains some things that we just sort of presume students will grok. i.e. we could just have text that says "check your directory" and links to a quick explainer on where files are in the directory structure and how to confirm you're in the right place. That would help students develop an intuition for this piece which I think otherwise makes the tutorials a bit fragile.

**However/Overall**
 - For me, the proposed changes aren't "idiomatic" enough to be the only motivation.
 - Consistency with the `pwd` commands across tutorials isn't a big deal and may be a red herring.
 - I acknowledge I may be overestimating the disruption that the additional subdirectory would cause in VS code. Setting the `cwd` in launch.json takes care of most of this, and we'll just need to add a reminder to students that if they open an integrated terminal that they need to `cd` into the subdirectory.

So, I'm alright with moving everything to a two-directory structure if you feel like the consistency will help elsewhere. You've also got more experience and time in the tutorials across the board and your intuition might prove to be right!

**Comment by awdeorio** [comment]
@jamesjuett You make two great points, which I've responded to by creating targeted issues.

> The pwd Cue and Consistency
> I don't think most students use this cue anyway. Those that understand it don't really need it, and those that don't understand it just execute pwd by rote in their own terminal. This at least is my experience interacting with students in office hours.

#33

> Either of those approaches (or the current pwd cue) might be better supported if we had a quick tutorial for terminal basics

#32

You bring up a point that I'll address shortly by adding a commit to #30:

> add a reminder to students that if they open an integrated terminal that they need to cd into the subdirectory.

I appreciate your flexibility on this.  I will move forward with the two-level directory approach using the `src` nomenclature.  I'll also address #32 and #33.  If after a test drive, the two-level structure turns out to be more trouble than its worth for VS Code students, I'm open to changing it back to a one-level structure.

**Comment by awdeorio** [comment]
TL;DR: I think I need to walk back my recommendation to move to a two-level directory structure and stick with the one-level structure.  I still firmly recommend project creation within the IDE tutorial.

I was wrong about two things in Visual Studio.  First, it is easy to create one-level directory structure.  There's an option for it I didn't know about.  Second, it is difficult to rename the inner folder (e.g., `p1-stats` -> `src`).  There's an option, but it renames a *reference* not the underlying file.

This undermines my plan to move to a two-level directory structure.

### Option 1: One-level `p1-stats/main.cpp`
This is the status quo.  It's easy in VS Code and Visual Studio.  Requires an extra step with Xcode (drag-and-drop).

In command line use cases, it's simpler to use a one-level directory structure.  VS Code users probably interact with the CLI more than Visual Studio or Xcode users.

A downside is that it's inconsistent with EECS 281, but I that's probably not a deal breaker.

### Option 2: Two-level `p1-stats/p1-stats/main.cpp`
Easy to create with Visual Studio, Xcode, and VS Code.  Consistent with Xcode and Visual Studio defaults.  Consistent with EECS 183 and EECS 281.

Risk of confusion with two directories of the same name, particularly with command line and VS Code.  Complicates the lives of VS Code users.

### Option 3: Two-level `p1-stats/src/main.cpp`
This is easy in VS Code and Xcode, but requires complicated gymnastics in Visual Studio.  Avoids confusion of two directories with the same name.  I used to think this was a good option, but I don't anymore.

### Option 4: Mixed
VS Code and maybe Visual Studio users could use a one-level directory structure, while Xcode users could use a two-level directory structure.

Risk of confusion, particularly around anything at the command line.
